virtcontainers/sandbox.go

```diff
@@ -1062,6 +1068,19 @@ func (s *Sandbox) stopVM() error {
 		s.Logger().Info("Not stopping VM")
 		return nil
 	}
+	pid := s.hypervisor.pid()
+
+	defer func() {
+		for i := 0; i < waitRetry; i++ {
+			if running, _ := isProcRunning(pid); !running {
+				return
+			}
+			syscall.Kill(pid, syscall.SIGKILL)
+			time.Sleep(waitVMExitTime * time.Millisecond)
+		}
+		networkLogger().WithField("pid", pid).
+			Warnf("NOTE: hypervisor not exited after wait %ds", (waitRetry*waitVMExitTime)/(time.Second/time.Millisecond))
+	}()
 
 	s.Logger().Info("Stopping VM")
 	return s.hypervisor.stopSandbox()
```

Comment:
@Ace-Tang 
 >  I update. Should we skip the check if s.hypervisor.stopSandbox() get error ?

good catch!  I think we shouldn't skip that check, instead you can modify the signature of this function: 
```go
(s *Sandbox) stopVM() (err error) {
``` 
Log the error if `s.hypervisor.stopSandbox()` fails: 
```go
if err = s.hypervisor.stopSandbox(); err != nil {
   // Log error
}
```
 and finally modify the defer function

```go
defer func() {
    if err == nil {
        return
    }
    for i := 0; i < waitRetry; i++ {
         if running, _ := isProcRunning(pid); !running {
               err = nil
               return
          }
          .......
     // convert this is an error
     // networkLogger().WithField("pid", pid)...
    return fmt.Errorf("hypervisor not exited after ......")
```

what do you think ?

Reply:
Even `s.hypervisor.stopSandbox()` return nil, the qemu may still living for a while, so I think `err
== nil`, the check is need, but if `s.hypervisor.stopSandbox()` return error, do we still need to check ? like

```
defer func() {
    if err != nil {
        // log error
        return 
    }
    for i := 0; i < waitRetry; i++ {
         if running, _ := isProcRunning(pid); !running {
               err = nil
               return
          }
....
     return fmt.Errorf("hypervisor not exited after ......")

```

Reply:
@Ace-Tang 

> Even s.hypervisor.stopSandbox() return nil, the qemu may still living for a while, so I think err == nil, the check is need,

does `qmp` not shutdown the vm at all and return nil? 
if so, then `err == nil` condition is not needed, the error should be logged and the hypervisor killed, I think

```go
defer func() {
    if err != nil {
        // just log the error
    }
    for i := 0; i < waitRetry; i++ {
         if running, _ := isProcRunning(pid); !running {
               err = nil
               return
          }
.....
```

Reply:
@Ace-Tang Yes, if its the hypervisor taking some time to exit, without qmp returning any error, then we need to just log the error in case we get one and wait for the hypervisor to stop in both the error/non-error case.

Reply:
> does qmp not shutdown the vm at all and return nil?

Yes, this is reason cause kernel panic, qemu still get vfio, and iommu try to unbind vf from vfio.
And Updated.